--- a/mc_forward.c
+++ b/mc_forward.c
@@ -332,7 +332,11 @@
 	struct hlist_head *rhead = NULL;
 	int is_management;
 	int passup = 0;
-
+	bool promisc;
+	
+	struct net_device *brdev = BR_INPUT_SKB_CB(skb)->brdev;
+	promisc = !!(brdev->flags & IFF_PROMISC);
+	
 	eh = eth_hdr(skb);
 	etype = ntohs(eh->h_proto);
 
@@ -403,7 +407,7 @@
 
 		if (passup) {
 			/*multicast router is enabled, passing up for routing*/
-			os_br_pass_frame_up(skb);
+			os_br_pass_frame_up(skb, promisc);
 		} else
 			kfree_skb(skb);
 		return 0;
@@ -434,7 +438,7 @@
 		mc_do_router_flood(mdb, rhead, skb);
 
 	if (passup)
-		os_br_pass_frame_up(skb);
+		os_br_pass_frame_up(skb, promisc);
 	else
 		kfree_skb(skb);
 
--- a/mc_osdep.h
+++ b/mc_osdep.h
@@ -22,9 +22,9 @@
 #include <br_private.h>
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 0, 0))
-static inline int os_br_pass_frame_up(struct sk_buff *skb)
+static inline int os_br_pass_frame_up(struct sk_buff *skb, bool promisc)
 {
-	return br_pass_frame_up(skb);
+	return br_pass_frame_up(skb, promisc);
 }
 #else
 static inline int os_br_pass_frame_up(struct sk_buff *skb)
@@ -189,7 +189,7 @@
 
 	dst = os_br_fdb_get((struct net_bridge *)br, eth_hdr(*skb)->h_dest);
 
-	if (dst && !dst->is_local)
+	if (dst && !test_bit(BR_FDB_LOCAL, &dst->flags))
 		return dst->dst;
 
 	return NULL;
--- a/mc_snooping.c
+++ b/mc_snooping.c
@@ -3442,6 +3442,18 @@
         return single_open(file, mc_proc_snooper_show, NULL);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,6,0)
+#define HAVE_PROC_OPS
+#endif
+
+#ifdef HAVE_PROC_OPS
+static const struct proc_ops mc_proc_snooper_fops = {
+  .proc_open = mc_proc_snooper_open,
+  .proc_read = seq_read,
+  .proc_lseek = seq_lseek,
+  .proc_release = single_release,
+};
+#else
 static const struct file_operations mc_proc_snooper_fops = {
         .owner          = THIS_MODULE,
         .open           = mc_proc_snooper_open,
@@ -3449,6 +3461,7 @@
         .llseek         = seq_lseek,
         .release        = single_release,
 };
+#endif
 
 /* mc_proc_create_snooper_entry
  *	create proc entry for information show
